src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs

```diff
@@ -63,6 +63,19 @@ public static Document GetRequiredDocument(this Solution solution, DocumentId do
     }
 
 #if !CODE_STYLE
+    public static SourceGeneratedDocument GetRequiredSourceGeneratedDocumentForAlreadyGeneratedId(this Solution solution, DocumentId documentId)
```

Comment:
I've used this helper method in a few spots in code actions, based on the paths that the current Razor tests hit, and it works but it doesn't fill me with confidence. I wonder how you feel about making `GetRequiredDocument` above, which is what everything calls, also return source generated documents if they're already been generated? Or to be even more strict, if they've been frozen?

It would certainly allow Razor to fall into the pit of success, but not sure if it allows too many others as well.

Reply:
eg, I just pushed [`a41720b` (#77587)](https://github.com/dotnet/roslyn/pull/77587/commits/a41720b40b12ff804692cdc60669bb77c221196d) because that spot wasn't hit on any of the Razor tests I was running yesterday, but what hit on the next one in line.

Finding each specific `GetRequiredDocument` call that needs to change is certainly possible, but also a potentially drawn out process.

Reply:
Found another! Without updating `GetRequiredDocument` Razor gets a yellow bar in completion, and no tooltips, because of this line:
https://github.com/dotnet/roslyn/blob/main/src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs#L229

Again, changing `GetRequiredDocument` to allow returning already generated documents would solve the issue. Could be limited to documents from the Razor source generator too.

Reply:
> The general exception is you call one of our GetRequiredDocumentAsync helpers which are allowed to return source generated documents

I guess I was trying to keep things like completion and code actions on the happy path of not causing generators to run, and only operating on frozen documents because we know those are the ones that a code fix actually changed. If this method was removed, and the callers used `GetRequiredDocumentAsync`, and processed all changed source generated documents, then you'd end up with the code action producing text edits for the generated documents due to changes that would simply happen anyway the next time the generator runs, if the source document updates.